### PR TITLE
chore: check for binaries

### PR DESCRIPTION
**Pull request by williamluke4**
[Related Issue](https://github.com/prisma/e2e-tests/issues/1729)

Checks for generated client files (`.prisma/client`) have been added for

- platforms-serverless
  - lambda
  - azure-functions-linux
  - gcp-functions
  - vercel-cli
  - vercel-node-builder
  - azure-functions-linux
  - firebase-functions (but napi also excluded below as it shows it is not working)
  - gcp-functions (but napi also excluded below as it shows it is not working)
- platforms
  - codesandbox
  - heroku  (but napi also excluded below as it shows it is not working)

---

New exclusions have been added for

- platforms-napi
  - heroku
- platforms-serverless-napi
  - platform: firebase-functions
  - platform: netlify-ci
  - ~~platform: gcp-functions~~

The later fix of these is tracked in https://github.com/prisma/e2e-tests/issues/1748

### CODE REVIEW

File: platforms-serverless/vercel-cli/test.sh [hunk 1]
```diff
@@ -2,6 +2,5 @@
 
 set -eu
 
-DEPLOYED_URL=$( tail -n 1 deployment-url.txt )
 
-npx ts-node ../../utils/fetch-retry-and-confirm-version.ts --url $DEPLOYED_URL --prisma-version $(sh ../../utils/prisma_version.sh) --binary-string ',"files":["index.js","package.json","query-engine-rhel-openssl-1.0.x","schema.prisma"]'
```
Comment (janpio) [inline platforms-serverless/vercel-cli/test.sh]: `DEPLOYED_URL` didn't work here?

File: platforms/codesandbox/deploy.ts [hunk 2]
```diff
@@ -98,6 +98,10 @@ async function main() {
       }
     })
     .reduce((files, file) => {
+      // Enables Node-API in Codesandbox 
+      if(file.filePath === 'prisma/.env' && process.env.PRISMA_FORCE_NAPI === 'true'){
+        file.content = file.content + `\nPRISMA_FORCE_NAPI=true`
+      }
```
Comment (janpio) [inline platforms/codesandbox/deploy.ts:104]: Ugh hack. Mabybe add a comment to the .env template we usually use that this file might be modified in deploy.ts

File: .github/workflows/test.yaml [hunk 3]
```diff
@@ -449,7 +449,7 @@ jobs:
         run: bash .github/slack/notify-failure.sh platforms ${{ matrix.platform }}
 
   platforms-napi:
-    needs: [start-time, platforms] # Makes these tests run after their binary counterparts
+    needs: [start-time] # Makes these tests run after their binary counterparts
```
Comment (janpio) [inline .github/workflows/test.yaml]:
```suggestion
    needs: [start-time]
```

File: .github/workflows/test.yaml [hunk 4]
```diff
@@ -593,9 +594,8 @@ jobs:
         run: bash .github/slack/notify-failure.sh platforms-serverless ${{ matrix.platform }}
 
   platforms-serverless-napi:
-    needs: [start-time, platforms-serverless] # Makes these tests run after their binary counterparts
+    needs: [start-time] # Makes these tests run after their binary counterparts
```
Comment (janpio) [inline .github/workflows/test.yaml]:
```suggestion
    needs: [start-time]
```

File: platforms-serverless/netlify-ci/functions/index.js [hunk 5]
```diff
@@ -39,10 +39,11 @@ exports.handler = async function (event, context, callback) {
   const tree = dirTree(process.env.LAMBDA_TASK_ROOT);
   console.dir(tree, { depth: null });
   */
- 
+
   // list all files in node_modules/.prisma/client
   const fs = require('fs')
-  const files = fs.readdirSync(process.env.LAMBDA_TASK_ROOT + "/src/node_modules/.prisma/client")
+  const path = require('path')
+  const files = fs.readdirSync(path.dirname(require.resolve('.prisma/client')))
```
Comment (janpio) [inline platforms-serverless/netlify-ci/functions/index.js]: Why this change? That code works in AWS Lambda generally and is used in other existing tests.

File: platforms-serverless/vercel-cli/test.sh [hunk 6]
```diff
@@ -2,4 +2,11 @@
...
 
-npx ts-node ../../utils/fetch-retry-and-confirm-version.ts --url https://e2e-vercel-api.vercel.app/api --prisma-version $(sh ../../utils/prisma_version.sh) --binary-string ',"files":["index.js","package.json","query-engine-rhel-openssl-1.0.x","schema.prisma"]'
+# checks whether PRISMA_FORCE_NAPI has length equal to zero
+if [[ -z "${PRISMA_FORCE_NAPI+x}" ]]; then
+  files=',"files":["index.js","package.json","query-engine-rhel-openssl-1.0.x","schema.prisma"]'
+else
+  files=',"files":["index.js","libquery_engine_napi-rhel-openssl-1.0.x.so.node","package.json","schema.prisma"]'
+fi
+
+npx ts-node ../../utils/fetch-retry-and-confirm-version.ts --url https://e2e-vercel-api.vercel.app/api --prisma-version $(sh ../../utils/prisma_version.sh) --binary-string $files
```
Comment (janpio) [inline platforms-serverless/vercel-cli/test.sh:13]: Can we also get the deployment URL to use here?

File: platforms-serverless/vercel-with-nextjs/run.sh [hunk 7]
```diff
@@ -8,5 +8,11 @@ export VERCEL_PROJECT_ID=$VERCEL_WITH_NEXTJS_PROJECT_ID
...
 echo "VERCEL_PROJECT_ID: $VERCEL_PROJECT_ID"
 echo "VERCEL_ORG_ID: $VERCEL_ORG_ID"
-yarn vercel --token=$VERCEL_TOKEN --prod --scope=prisma --confirm
+
+# checks whether PRISMA_FORCE_NAPI has length equal to zero
+if [[ -z "${PRISMA_FORCE_NAPI+x}" ]]; then
+  yarn vercel --token=$VERCEL_TOKEN --prod --scope=prisma --confirm
+else
+  yarn vercel --token=$VERCEL_TOKEN --env PRISMA_FORCE_NAPI=true --build-env PRISMA_FORCE_NAPI=true --prod --scope=prisma --confirm
+fi
```
Comment (janpio) [inline platforms-serverless/vercel-with-nextjs/run.sh]: This is missing the check part in the code.

File: platforms-serverless/vercel-with-redwood/run.sh [hunk 8]
```diff
@@ -10,5 +10,11 @@ export FORCE_RUNTIME_TAG=canary
...
 echo "VERCEL_ORG_ID: $VERCEL_ORG_ID"
 echo "FORCE_RUNTIME_TAG $FORCE_RUNTIME_TAG"
-yarn vercel --token=$VERCEL_TOKEN --prod --scope=prisma --confirm --force
+
+
+if [[ -z "${PRISMA_FORCE_NAPI+x}" ]]; then
+  yarn vercel --token=$VERCEL_TOKEN --prod --scope=prisma --confirm --force
+else
+  yarn vercel --token=$VERCEL_TOKEN --env PRISMA_FORCE_NAPI=true --build-env PRISMA_FORCE_NAPI=true --prod --scope=prisma --confirm --force
+fi
```
Comment (janpio) [inline platforms-serverless/vercel-with-redwood/run.sh]: This is missing the check part in the code.

File: platforms-serverless/vercel-cli/test.sh (hunk 1 above)
Comment (williamluke4) [reply platforms-serverless/vercel-cli/test.sh]: Nope, have no idea why

File: platforms-serverless/netlify-ci/functions/index.js (hunk 5 above)
Comment (williamluke4) [reply platforms-serverless/netlify-ci/functions/index.js]: Please see https://prisma-company.slack.com/archives/CRGDU19K6/p1623240841039500

File: platforms-serverless/vercel-cli/test.sh (hunk 6 above)
Comment (williamluke4) [reply platforms-serverless/vercel-cli/test.sh:13]: I tried using it but for some reason it did not work

File: platforms-serverless/vercel-with-redwood/run.sh (hunk 8 above)
Comment (williamluke4) [reply platforms-serverless/vercel-with-redwood/run.sh]: Added

File: platforms-serverless/vercel-with-nextjs/index.test.js [hunk 9]
```diff
@@ -1,8 +1,16 @@
...
-const endpoint = 'https://e2e-vercel-with-nextjs.vercel.app/api'
+// const endpoint = 'https://e2e-vercel-with-nextjs.vercel.app/api'
 // const endpoint = 'http://localhost:3001/api'
 
+function getDeploymentURL(){
+  const data = fs.readFileSync('./deployment-url.txt', {encoding: 'utf8'})
+  let lines = data.trim().split("\n")
+  return lines[lines.length - 1].trim()
+}
+const endpoint = getDeploymentURL() + '/api/graphql'
```
Comment (janpio) [inline platforms-serverless/vercel-with-nextjs/index.test.js]:
```suggestion
const endpoint = getDeploymentURL() + '/api'
```
probably

Comment (williamluke4) [reply platforms-serverless/vercel-with-nextjs/index.test.js]: yep :see_no_evil:

File: platforms-serverless/vercel-with-nextjs/pages/api/files.js [hunk 10]
```diff
@@ -12,5 +12,7 @@ export default async (req, res) => {
   }
   res.status(200).json({
     files,
+    napi: process.env.PRISMA_FORCE_NAPI,
+    connection: process.env.NEXTJS_WITH_VERCEL_PG_URL,
```
Comment (janpio) [inline platforms-serverless/vercel-with-nextjs/pages/api/files.js]: Weird, I get this back:

```
{"files":["index.js","package.json","query-engine-rhel-openssl-1.0.x","schema.prisma"],"connection":"postgresql://prisma:qd...eu-central-1.rds.amazonaws.com:5432/platform-vercel-nextjs?schema=public"}
```

https://e2e-vercel-with-nextjs-by6g8z242-prisma.vercel.app/api/files

Why is `napi` just missing!?

Comment (janpio) [reply platforms-serverless/vercel-with-nextjs/pages/api/files.js]: Seems the database URL is set via the UI: https://vercel.com/prisma/e2e-vercel-with-nextjs/settings/environment-variables

image

Comment (janpio) [reply platforms-serverless/vercel-with-nextjs/pages/api/files.js]: Oh right, it _has_ to be empty as you are not setting it any more with the same commit: [`8ac579c` (#1764)](https://github.com/prisma/e2e-tests/pull/1764/commits/8ac579c93d28c70f5e4d95f0e026d785aaf23b7b#diff-e88497bc11548c0857682529493a6398a1d7dfb8e717f07add73147980733252L16)

But why it is really _not listed_ in the response I do not understand.

File: platforms-serverless/vercel-with-nextjs/prisma/schema-with-napi.prisma [hunk 11]
```diff
@@ -0,0 +1,16 @@
+generator client {
+  provider = "prisma-client-js"
+  previewFeatures = ["nApi"]
```
Comment (janpio) [inline platforms-serverless/vercel-with-nextjs/prisma/schema-with-napi.prisma]: https://vercel.com/prisma/e2e-vercel-with-nextjs/933UQzXKzDeUksF75HZknGbxswzS/source?f=src%2Fprisma%2Fschema.prisma shows that this schema is indeed deployed.

But the Client still gets the Query Engine binary, and not the library: 

```
{"files":["index.js","package.json","query-engine-rhel-openssl-1.0.x","schema.prisma"] ...}
```
https://e2e-vercel-with-nextjs-by6g8z242-prisma.vercel.app/api/files

Comment (janpio) [reply platforms-serverless/vercel-with-nextjs/prisma/schema-with-napi.prisma]: Maaayyyyybe there is some special Vercel logic which engine to deploy that does not consider Napi properly? Or something else goes wrong during the build process? 
(Or do we actually upload the local build, which is not using `rhel` but `debian` and then something goes wrong?)

So many possibilities.

File: .github/workflows/test.yaml [hunk 12]
```diff
@@ -460,8 +460,10 @@ jobs:
           - aws-graviton
           - codesandbox
         os: [ubuntu-latest] #, windows-latest, macos-latest]
+        exclude:
+          - platform: heroku # https://github.com/prisma/e2e-tests/issues/1773
```
Comment (janpio) [inline .github/workflows/test.yaml]:
```suggestion
          - platform: heroku # https://github.com/prisma/e2e-tests/issues/1748
```

File: platforms/codesandbox/test.sh [hunk 13]
```diff
@@ -7,8 +7,16 @@ sandbox_id=$(cat sandbox_id)
 url="https://$sandbox_id.sse.codesandbox.io"
 
 prisma_version="$(cat ../../.github/prisma-version.txt)"
+# }'
```
Comment (janpio) [inline platforms/codesandbox/test.sh]:
```suggestion
```

File: platforms-serverless/netlify-ci/test.sh [hunk 14]
```diff
@@ -2,5 +2,6 @@
 
 set -eu
 
+# TODO This is not dynamic based on Napi env var, so should not even work... wtf?
```
Comment (williamluke4) [inline platforms-serverless/netlify-ci/test.sh:5]: That's because we don't/can't set the ENV VAR in the CI. Thus it will always use the binary

File: platforms-serverless/vercel-with-nextjs/pages/api/files.js (hunk 10 above)
Comment (williamluke4) [reply platforms-serverless/vercel-with-nextjs/pages/api/files.js]: This was because we were not using `vercel --force` which clears the build cache

File: platforms-serverless/gcp-functions/run.sh [hunk 15]
```diff
@@ -10,7 +10,7 @@ if [[ -z "${PRISMA_FORCE_NAPI+x}" ]]; then
   # use the default schema at prisma/schema.prisma file
   true
 else
-  mv ./prisma/schema-with-napi.prisma ./prisma/schema.prisma
```
Comment (janpio) [inline platforms-serverless/gcp-functions/run.sh]: what triggered this? just curious - moving would make sure the other file is gone.

Comment (williamluke4) [reply platforms-serverless/gcp-functions/run.sh]: The retry logic then fails as it can't move a file thats already been moved

Comment (janpio) [reply platforms-serverless/gcp-functions/run.sh]: Ah, commit message: `@williamluke4 change mv to cp to prevent retries from failing`
And this is of course the case because we retry multiple times with the same checkout.
But doesn't this also lead to the binary test to be a library test the next run? (So we actually would need to backup the file, and write it back after the test?)

Comment (williamluke4) [reply platforms-serverless/gcp-functions/run.sh]: No, because the binary and Node-API tests are in different jobs (i.e clean)

Comment (janpio) [reply platforms-serverless/gcp-functions/run.sh]: Ah true, we changed that.
